js/src/common/states/PaginatedListState.ts

```diff
@@ -0,0 +1,171 @@
+import * as Mithril from 'mithril';
+
+import Model from '../Model';
+
+export interface Page<TModel> {
+  number: number;
+  items: TModel[];
+
+  // // maybe we can put page links here?
+  // links: {[name: string]: string},
+  // // or
+  // previousPage?: string;
+  // nextPage?: string;
+}
+
+export interface PaginationLocation {
+  page: number;
+  startIndex?: number;
+  endIndex?: number;
+}
+
+export default abstract class PaginatedListState<T extends Model> {
+  // TODO:
+  //  - what does this page track? last page request? last page that is loaded
+  //      (e.g. pg 7 if u have 3-7 loaded)
+  //  - When do we set it?
+  protected location!: PaginationLocation;
+
+  // TODO: might want interface so pagination can work w/o model-specific stuff
+  protected type: string;
```

Comment:
Well, I suppose it could be any JSON:API resource. Frontend models represent those (not necessarily a db model), so requiring a model should be fine.